### ISSUE TITLE
unknown type name error in latest sndlib

### ISSUE THREAD

**Issue by Teteros** [kmatheussen/radium#1151]
RADIUM_QT_VERSION=5 make packages
RADIUM_QT_VERSION=5 BUILDTYPE=RELEASE ./build_linux.sh
```
gcc -mfpmath=sse -msse2 -c -DHAVE_CONFIG_H -I.  -g  -Og   -O3 -fPIC     sndlib2xen.c
In file included from sndlib2xen.c:33:
vct.h:67:12: error: unknown type name ‘Xen’
 MUS_EXPORT Xen mus_array_to_list(mus_float_t *arr, mus_long_t i, mus_long_t len);
            ^~~
vct.h:74:12: error: unknown type name ‘Xen’
 MUS_EXPORT Xen xen_list_to_vct(Xen lst);
            ^~~
vct.h:74:32: error: unknown type name ‘Xen’
 MUS_EXPORT Xen xen_list_to_vct(Xen lst);
                                ^~~
vct.h:75:28: error: unknown type name ‘Xen’
 MUS_EXPORT vct *xen_to_vct(Xen arg);
                            ^~~
vct.h:76:12: error: unknown type name ‘Xen’
 MUS_EXPORT Xen xen_make_vct(mus_long_t len, mus_float_t *data);
            ^~~
vct.h:77:12: error: unknown type name ‘Xen’
 MUS_EXPORT Xen xen_make_vct_wrapper(mus_long_t len, mus_float_t *data);
            ^~~
vct.h:78:12: error: unknown type name ‘Xen’
 MUS_EXPORT Xen g_vct_peak(Xen obj);
            ^~~
vct.h:78:27: error: unknown type name ‘Xen’
 MUS_EXPORT Xen g_vct_peak(Xen obj);
                           ^~~
vct.h:94:30: error: unknown type name ‘Xen’
   MUS_EXPORT bool mus_is_vct(Xen obj);
                              ^~~
vct.h:95:14: error: unknown type name ‘Xen’
   MUS_EXPORT Xen vct_to_xen(vct *v);
              ^~~
make[1]: *** [makefile:49: sndlib2xen.o] Error 1
```
repo version: 5.8.6 + 04ae9149432f597e0047ada0194db8e4f1f0f746
gcc --version: `gcc (GCC) 8.2.1 20180831` (Arch Linux)

sndlib compiles correctly if I compile manually:

```bash
cd bin/packages
tar xvzf sndlib.tar.gz
cd sndlib
patch -p0 <../sndlib.patch
./configure
make
```

Output:

```
$ ./configure
checking build system type... x86_64-unknown-linux-gnu
checking host system type... x86_64-unknown-linux-gnu
checking for gcc... gcc
checking whether the C compiler works... yes
checking for C compiler default output file name... a.out
checking for suffix of executables...
checking whether we are cross compiling... no
checking for suffix of object files... o
checking whether we are using the GNU C compiler... yes
checking whether gcc accepts -g... yes
checking for gcc option to accept ISO C89... none needed
checking for a BSD-compatible install... /usr/bin/install -c
checking how to run the C preprocessor... gcc -E
checking for grep that handles long lines and -e... /usr/bin/grep
checking for egrep... /usr/bin/grep -E
checking for ANSI C header files... yes
checking for sys/types.h... yes
checking for sys/stat.h... yes
checking for stdlib.h... yes
checking for string.h... yes
checking for memory.h... yes
checking for strings.h... yes
checking for inttypes.h... yes
checking for stdint.h... yes
checking for unistd.h... yes
checking whether byte ordering is bigendian... no
checking size of void *... 8
checking for pkg-config... /usr/bin/pkg-config
checking for gsl... yes
checking for audio system... ALSA
configure: creating ./config.status
config.status: creating makefile
config.status: creating sndlib.pc
config.status: creating sndins/Makefile
config.status: creating sndlib-config
config.status: creating unix-config.h
$ make
gcc -c -DHAVE_CONFIG_H -I. -g -O2 -fPIC     sndlib2xen.c
gcc -c -DHAVE_CONFIG_H -I. -g -O2 -fPIC     clm2xen.c
gcc -c -DHAVE_CONFIG_H -I. -g -O2 -fPIC     s7.c
s7.c: In function ‘s7_make_keyword’:
s7.c:9209:3: warning: ‘memcpy’ specified size 18446744073709551614 exceeds maximum object size 9223372036854775807 [-Wstringop-overflow=]
   memcpy((void *)(name + 1), (void *)key, slen);     /* gcc 8.1 error here is incorrect */
   ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
gcc  headers.o audio.o io.o sound.o xen.o vct.o clm.o sndlib2xen.o clm2xen.o s7.o -o libsndlib.so -shared  -lasound -lgsl -lgslcblas -lm    -lm -ldl
ar -rc libsndlib.a headers.o audio.o io.o sound.o xen.o vct.o clm.o sndlib2xen.o clm2xen.o s7.o
: libsndlib.a
```

It might be something in Makefile.Qt

**Comment by kmatheussen** [comment]
Ouch, this error should have popped up on the build machine too. Seems like latest version of sndlib (with custom patch) wasn't included in the binaries.

The build fix is simple though. Just include #include "xen.h" in the top of vct.h.

**Comment by Teteros** [comment]
Thanks for 5.8.7 which includes the newest sndlib and build fix.

Two issues with compile remain (clean git repo clone):

- Build stops at `sndlib_built` step in Makefile.Qt

```
[...]
packages successfully built. Now run ./build_linux.sh -j7
$ RADIUM_QT_VERSION=5 BUILDTYPE=RELEASE ./build_linux.sh -j`nproc`
RADIUM_QT_VERSION=5
BUILDTYPE=RELEASE
INCLUDE_FAUSTDEV=jadda
INCLUDE_PDDEV=jadda
INCLUDE_FAUSTDEV=jadda
INCLUDE_PDDEV=jadda
  Non-linux build detected. Deleting binaries.
echo linux
linux
if [ RELEASE = DEBUG ] ; then \
        echo "-Werror -Wall -g -fno-optimize-sibling-calls -fsanitize=undefined -fno-sanitize-recover  -fsanitize=address -fno-common -DDISABLE_BDWGC " >buildtype.opt ; \
else \
        echo "-DRELEASE -Wall -g -O2 -mfpmath=sse -msse2  -DNDEBUG" >buildtype.opt ; \
fi
if [ RELEASE = DEBUG ] ; then \
        echo "-DNOPAUSEPLAY -DCOMPILING_RADIUM -DUSE_WEAK_JACK -DGUIISQT -DUSE_GFX_OP_QUEUE -DFULL_VERSION=1 -DIS_LITTLE_ENDIAN=1 -DUSE_OPENGL=1 -DUSE_QT_VISUAL=1 -DUSE_GTK_VISUAL=0 -DUSE_QT_REQTYPE=1 -DUSE_GTK_REQTYPE=0 -DUSE_QT_MENU=1 -DUSE_GTK_MENU=0" >flagopts.opt ; \
else \
        echo "-DNOPAUSEPLAY -DCOMPILING_RADIUM -DUSE_WEAK_JACK -DGUIISQT -DUSE_GFX_OP_QUEUE -DFULL_VERSION=1 -DIS_LITTLE_ENDIAN=1 -DUSE_OPENGL=1 -DUSE_QT_VISUAL=1 -DUSE_GTK_VISUAL=0 -DUSE_QT_REQTYPE=1 -DUSE_GTK_REQTYPE=0 -DUSE_QT_MENU=1 -DUSE_GTK_MENU=0 -DRELEASE" >flagopts.opt ; \
fi
cd bin && /usr/bin/python2 generate_keyboard_sub_ids.py >../common/keyboard_sub_ids.h
bash check_dependencies.sh `./find_python_path.sh` "`/home/tete/code/aursrc/radium/src/radium-5.8.7/find_moc_and_uic_paths.sh moc`" "`/home/tete/code/aursrc/radium/src/radium-5.8.7/find_moc_and_uic_paths.sh uic`" test_build
/usr/bin/python2 ../bin/protoconfparser.py radium_proc.h
cd api/ ; /usr/bin/python2 ../bin/protoconfparser.py radium_proc.h ; cd ..
cd Qt/ && "`/home/tete/code/aursrc/radium/src/radium-5.8.7/find_moc_and_uic_paths.sh moc`" Qt_PluginWidget.h >mQt_PluginWidget.cpp
cd Qt/ && "`/home/tete/code/aursrc/radium/src/radium-5.8.7/find_moc_and_uic_paths.sh uic`" qt4_vst_paths_widget.ui | sed s/protected/public/ | sed s/QSpinBox/MyQSpinBox/ | sed s/QDoubleSpinBox/FocusSnifferQDoubleSpinBox/ | sed s:\#include\ \<QtGui/FocusSnifferQDoubleSpinBox\>://\ \ qspingox.h: | sed s:\#include\ \<QWidgets/FocusSnifferQDoubleSpinBox\>://\ \ qspingox.h: | sed s:\#include\ \<QtGui/MyQSpinBox\>://\ \ qspingox.h: | sed s:\#include\ \<QtWidgets/MyQSpinBox\>://\ \ qspingox.h: | sed s/QLineEdit/FocusSnifferQLineEdit/ | sed s:\#include\ \<QtWidgets/FocusSnifferQLineEdit\>://\ \ qlineedit.h: > Qt_vst_paths_widget.h
cd bin/packages && rm -fr s7 && tar xvzf s7.tar.gz && touch s7/s7.c
cd bin/packages && rm -fr sndlib && tar xvzf sndlib.tar.gz && cd sndlib && patch -p0 <../sndlib.patch && touch clm.h
make: *** No rule to make target 'bin/packages/sndlib/clm.c', needed by 'sndlib_built'.  Stop.
make: *** Waiting for unfinished jobs....
```

I unpacked sndlib manually before building packages with `bin/packages/sndlib.tar.gz -C bin/packages`
and this got through.

- Linker complaints about undefined reference to `QsciScintilla::wheelEvent(QWheelEvent*)'` in `Qt_instruments.o`

```
g++ `cat buildtype.opt` Qt_instruments.o Qt_visual.o GTK_visual.o Qt_Main.o Qt_endprogram.o Qt_EventReceiver.o Qt_colors.o Qt_Menues.o Qt_Fonts.o Qt_ReqType.o Qt_PopupMenu.o Qt_Bs_edit.o Qt_PluginWidget.o Qt_SliderPainter.o Qt_memory.o Qt_path_resolver.o Qt_settings.o Qt_disk.o Qt_MainWindow.o Qt_Time.o Qt_sequencer.o GTK_ReqType.o GTK_PopupMenu.o Qt_soundfilesaver_widget_callbacks.o Qt_comment_dialog.o Qt_song_properties.o Qt_preferences_callbacks.o Qt_tools_callbacks.o Qt_pianorollheader_callbacks.o Qt_Error.o Qt_progresswindow.o Qt_check_for_updates.o Rational.o Qt_AutoBackups.o flowlayout.o scheme.o bin/packages/s7/s7_linux_RELEASE.o sndlib_objectfiles/headers.o sndlib_objectfiles/audio.o sndlib_objectfiles/io.o sndlib_objectfiles/sound.o sndlib_objectfiles/xen.o sndlib_objectfiles/vct.o sndlib_objectfiles/clm.o sndlib_objectfiles/sndlib2xen.o sndlib_objectfiles/clm2xen.o s7webserver.o radium_s7_wrap.o qhttpconnection.o moc_qhttpconnection.o qhttprequest.o moc_qhttprequest.o qhttpresponse.o moc_qhttpresponse.o qhttpserver.o moc_qhttpserver.o http_parser.o windows.o window_config.o list.o vector.o hashmap.o Dynvec.o song.o blocks.o block_insert.o block_split.o block_delete.o block_properties.o tracks.o localzooms.o control.o lines.o font.o track_insert.o track_onoff.o settings.o mmd2load.o notes.o notes_legalize.o pitches.o wblocks.o wtracks.o sliders.o gfx_wblocks.o gfx_wblocks_reltempo.o common.o gfx_wtext.o gfx_text.o gfx_lines.o gfx_point.o gfx_op_queue.o eventreciever.o scancodes.o reallines.o notestext.o trackreallines2.o signaturetext.o lpbtext.o bpmtext.o swingtext.o centtext.o chancetext.o veltext.o fxtext.o data_as_text.o clipboard_range.o clipboard_range_calc.o clipboard_range_copy.o clipboard_range_paste.o clipboard_range_cut.o transpose.o backwards.o invert.o glissando.o pitchexpand.o expand.o clipboard_track_copy.o clipboard_track_paste.o clipboard_track_cut.o clipboard_tempos_copy.o clipboard_localzooms.o clipboard_block_copy.o clipboard_block_paste.o quantitize.o tbox.o area.o range.o nodelines.o debug.o memory.o validatemem.o placement.o t_gc.o cursor.o cursor_updown.o velocities.o pixmap.o scroll.o blts.o realline_calc.o gfx_subtrack.o LPB.o Signature.o Beats.o gfx_wtrackheaders.o gfx_wtrackheader_volpan.o gfx_slider.o reallines_insert.o gfx_shrink.o gfx_shrink_t.o gfx_request.o instruments.o patch.o undo_audio_patch_addremove.o fxlines.o fxlines_legalize.o seqtrack.o seqtrack_automation.o song_tempo_automation.o seqblock_automation.o scroll_play.o crashreporter.o crashreporter_posix.o crashreporter_windows.o error.o  threading.o Semaphores.o stacktoucher.o reltempo.o temponodes.o tempos.o time.o temponodes_legalize.o swing.o Ptask2Mtask.o player.o scheduler.o scheduler_seqtracks.o scheduler_notes.o scheduler_pitches.o scheduler_velocities.o scheduler_fxs.o scheduler_reallines.o scheduler_LPB.o scheduler_Beats.o scheduler_Signature.o playerclass.o player_startstop.o player_pause.o  disk.o disk_fxs.o disk_wblock.o disk_localzooms.o disk_track.o disk_fx.o disk_fxnodelines.o disk_wtrack.o disk_temponodes.o disk_tempos.o disk_song.o disk_velocities.o disk_block.o disk_placement.o disk_stops.o disk_playlist.o disk_root.o disk_notes.o disk_lpbs.o disk_signatures.o disk_swing.o disk_windows.o disk_warea.o disk_save.o disk_load.o disk_saveload_blocktrack.o disk_instrument.o disk_patches.o disk_slider.o undo.o undo_notes.o undo_fxs.o undo_temponodes.o undo_tempos.o undo_swings.o undo_lpbs.o undo_signatures.o undo_notesandfxs.o undo_reallines.o undo_tracks.o undo_range.o undo_blocks.o undo_trackheader.o undo_reltempomax.o undo_maintempos.o undo_block_insertdelete.o undo_block_mergesplit.o undo_reltemposlider.o undo_sequencer.o undo_song_tempo_automation.o undo_patchvoice.o undo_patchname.o undo_chip_position.o radium_wrap.o api_common.o api_keyplay.o api_keyplayedit.o api_navigate.o api_disk.o api_noteedit.o api_velocities.o api_pitches.o api_timing.o api_support.o wrapfunclist.o api_trackonoff.o api_zoom.o api_notemanipulate.o api_play.o api_clipboard.o api_undo.o api_various.o api_instruments.o api_soundpluginregistry.o api_requesters.o api_mouse.o api_seqtracks.o api_gui.o posix_settings.o posix_Player.o  X11_keyboard.o X11_disk.o X11_Qtstuff.o   disk_midi_fx.o disk_midi_i_plugin.o midi_fx.o midi_i_plugin.o midi_playfromstart.o midi_i_input.o midi_menues.o zita_rev.o stk_flute.o stk_bowed.o stk_blow_bottle.o stk_bass.o stk_blow_hole.o stk_brass.o stk_clarinet.o stk_flute_stk.o stk_glass_harmonica.o stk_harpsi.o stk_modal_bar.o stk_NLF_eks.o stk_NLF_fm.o stk_piano.o stk_saxophony.o stk_sitar.o stk_tibetan_bowl.o stk_tuned_bar.o stk_uni_bar.o stk_voice_form.o faust_tapiir.o faust_system_eq.o faust_system_tremolo.o faust_system_lowpass.o faust_system_highpass.o faust_system_lowshelf.o faust_system_highshelf.o faust_multibandcomp.o audio_instrument.o SoundProducer.o Jack_plugin.o find_jack_library.o Faust_plugins.o Bus_plugins.o Timeskew_plugin.o Patchbay_plugin.o MidiMessages_plugin.o VST_plugins.o Ladspa_plugins.o Sampler_plugin.o SampleRecorder.o SampleReader.o Seqtrack_plugin.o FluidSynth_plugin.o Pd_plugin.o Modulator_plugin.o SoundPluginRegistry.o SoundPlugin.o AudioMeterPeaks.o Peaks.o Mixer.o get_windows_commandlines.o Envelope.o Resampler.o SoundFonts.o Smooth.o Presets.o undo_audio_effect.o undo_connection_enabled.o undo_audio_connection_gain.o undo_pd_controllers.o undo_plugin_state.o SoundfileSaver.o system_compressor_wrapper.o KillJackd.o  QM_view.o QM_chip.o QM_MixerWidget.o undo_mixer.o undo_mixer_connections.o GfxElements.o T2.o Render.o Widget.o  -o bin/radium_linux.bin -g -lm bin/packages/Visualization-Library-master/src/vlVG/lib/libVLVG.a bin/packages/Visualization-Library-master/src/vlGraphics/lib/libVLGraphics.a bin/packages/Visualization-Library-master/src/vlCore/lib/libVLCore.a -lfreetype  -lGL -lGLU bin/packages/QScintilla_gpl-2.10.1/Qt4Qt5/libqscintilla2_qt5.a bin/packages/faust2/compiler/libfaust.a -luuid [...REDACTED LLVM4 .a to keep it short...] -lcrypto -lncurses bin/packages/libpd-master/libs/libpds.a pluginhost/Builds/Linux/build/libMyPluginHost.a -lasound -pthread -lrt -lX11 -ldl -lXext  -llrdf bin/packages/gc-7.4.4/.libs/libgc.a  -L/usr/lib -lpython2.7 bin/packages/libgig/src/.libs/libgig.a bin/packages/fluidsynth-1.1.6/src/.libs/libfluidsynth.a -ldbus-1  -lsndfile  -lsamplerate  -lQt5X11Extras -lQt5Gui -lQt5Core  -lglib-2.0  -lxcb -lxcb-keysyms -Wl,-Bstatic -lbfd -Wl,-Bdynamic -lz -liberty -lutil -lgmp -lmpfr -lmpc -ljack -ldl /usr/lib/libQt5UiTools.a `pkg-config --libs Qt5Gui --libs Qt5OpenGL --libs Qt5Network --libs Qt5Widgets --libs Qt5WebKitWidgets` -fPIC
/usr/bin/ld: Qt_instruments.o: in function `FocusSnifferQsciScintilla::wheelEvent(QWheelEvent*)':
/home/tete/code/aursrc/radium/src/radium-5.8.7/Qt/FocusSniffers.h:145: undefined reference to `QsciScintilla::wheelEvent(QWheelEvent*)'
collect2: error: ld returned 1 exit status
make: *** [Makefile:584: bin/radium_linux.bin] Error 1
```

I'm on qt5-base 5.11.1 and qscintilla-qt5 2.10.7

**Comment by kmatheussen** [comment]
Is bin/packages/QScintilla_gpl-2.10.1/Qt4Qt5/libqscintilla2_qt5.a empty?

**Comment by Teteros** [comment]
`test -e bin/packages/QScintilla_gpl-2.10.1/Qt4Qt5/libqscintilla2_qt5.a && echo Exists`
Exists

**Comment by kmatheussen** [comment]
Is it empty? Is the size 0?
Even if it larger than 0, you should try rebuilding it. It could have been
interrupted during build, for some reason.

**Comment by kmatheussen** [comment]
[kjetil@localhost radium]$ ls -la
bin/packages/QScintilla_gpl-2.10.1/Qt4Qt5/libqscintilla2_qt5.a
-rw-rw-r--. 1 kjetil kjetil 7082690 Aug 31 09:00
bin/packages/QScintilla_gpl-2.10.1/Qt4Qt5/libqscintilla2_qt5.a

**Comment by kmatheussen** [comment]
Another possibility is that the compiler uses header files from the system
version of qscintilla.
You could try this patch:

kjetil@localhost radium]$ git diff
diff --git a/Qt/FocusSniffers.h b/Qt/FocusSniffers.h
index 39da025..5aec3f7 100644
--- a/Qt/FocusSniffers.h
+++ b/Qt/FocusSniffers.h
@@ -88,8 +88,8 @@ MakeFocusSnifferClass(QLineEdit);
 #  pragma GCC diagnostic push
 #  pragma GCC diagnostic ignored "-Wsuggest-override"
 #endif
-#include <Qsci/qscilexer.h>
-#include <Qsci/qsciscintilla.h>  // <--- Much trouble. Any qt4
alternatives?
+#include "../bin/packages/QScintilla_gpl-2.10.1/Qt4Qt5/Qsci/qscilexer.h"
+#include
"../bin/packages/QScintilla_gpl-2.10.1/Qt4Qt5/Qsci/qsciscintilla.h"  //
<--- Much trouble. Any qt4 alternatives?
 #if __GNUC__ >= 5
 #  pragma GCC diagnostic pop
 #endif
diff --git a/Qt/Qt_faust_plugin_widget_callbacks.h
b/Qt/Qt_faust_plugin_widget_callbacks.h
index f65f8f5..5268fd3 100644
--- a/Qt/Qt_faust_plugin_widget_callbacks.h
+++ b/Qt/Qt_faust_plugin_widget_callbacks.h
@@ -44,8 +44,8 @@ Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA
 02111-1307, USA. */
 #  pragma GCC diagnostic push
 #  pragma GCC diagnostic ignored "-Wsuggest-override"
 #endif
-#include <Qsci/qscilexerjava.h>
-#include <Qsci/qscilexercpp.h>
+#include
"../bin/packages/QScintilla_gpl-2.10.1/Qt4Qt5/Qsci/qscilexerjava.h"
+#include "../bin/packages/QScintilla_gpl-2.10.1/Qt4Qt5/Qsci/qscilexercpp.h"
 #if __GNUC__ >= 5
 #  pragma GCC diagnostic pop
 #endif

**Comment by Teteros** [comment]
Thank you! It looks like it was indeed system instead of radium qscintilla headers being used which are breaking.

Closing this issue as sndbuild fix in a57b5128784b80a55a2307e42044b78f1b7c8e12 solves the first problem.

**Comment by kmatheussen** [comment]
@Teteros can you look at this one? https://github.com/kmatheussen/radium/issues/1153
